src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Performance/AvoidPotentiallyExpensiveCallWhenLogging.cs

```diff
@@ -0,0 +1,333 @@
+        private const int LogLevelInformation = 2;
+        private const int LogLevelWarning = 3;
+        private const int LogLevelError = 4;
+        private const int LogLevelCritical = 5;
+        private const int LogLevelPassedAsParameter = int.MinValue;
+
+        private static readonly DiagnosticDescriptor Rule = DiagnosticDescriptorHelper.Create(
+            RuleId,
+            CreateLocalizableResourceString(nameof(AvoidPotentiallyExpensiveCallWhenLoggingTitle)),
+            CreateLocalizableResourceString(nameof(AvoidPotentiallyExpensiveCallWhenLoggingMessage)),
+            DiagnosticCategory.Performance,
+            RuleLevel.IdeSuggestion,
+            CreateLocalizableResourceString(nameof(AvoidPotentiallyExpensiveCallWhenLoggingDescription)),
+            isPortedFxCopRule: false,
+            isDataflowRule: false);
+
+        public sealed override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics { get; } = ImmutableArray.Create(Rule);
+
+        public sealed override void Initialize(AnalysisContext context)
+        {
+            context.EnableConcurrentExecution();
+            context.ConfigureGeneratedCodeAnalysis(GeneratedCodeAnalysisFlags.None);
+            context.RegisterCompilationStartAction(OnCompilationStart);
+        }
+
+        private void OnCompilationStart(CompilationStartAnalysisContext context)
+        {
+            if (!RequiredSymbols.TryGetSymbols(context.Compilation, out var symbols))
+            {
+                return;
+            }
+
+            context.RegisterOperationAction(AnalyzeInvocation, OperationKind.Invocation);
+
+            void AnalyzeInvocation(OperationAnalysisContext context)
+            {
+                var invocation = (IInvocationOperation)context.Operation;
+
+                // Check if invocation is a logging invocation and capture the log level (either as IOperation or as int, depending if it is dynamic or not).
+                // Use these to check if the logging invocation is guarded by 'ILogger.IsEnabled' and bail out if it is.
+                if (!symbols.TryGetLogLevel(invocation, out var logLevelArgumentOperation, out var logLevel) ||
+                    symbols.IsGuardedByIsEnabled(currentOperation: invocation, logInvocation: invocation, logLevel, logLevelArgumentOperation))
+                {
+                    return;
+                }
+
+                var arguments = invocation.Arguments.Skip(invocation.IsExtensionMethodAndHasNoInstance() ? 1 : 0);
+
+                // At this stage we have a logging invocation that is not guarded.
+                // Check each argument if it is potentially expensive.
+                foreach (var argument in arguments)
+                {
+                    // Check the argument value after conversions to prevent noise (e.g. implicit conversions from null or from int to EventId).
+                    if (IsPotentiallyExpensive(argument.Value.WalkDownConversion()))
```

Comment:
This is a bit of a deviation from the description in the runtime issue, but without this, there would be a lot of noise.
The following lines would be flagged:

```C#
void M(ILogger logger, EventId eventId, Exception exception, Func<object, Exception, string> formatter)
{
    logger.Log(LogLevel.Trace, eventId, null, exception, formatter);
    logger.Log(LogLevel.Trace, 0, "literal", exception, formatter);
}
```